#!/bin/bash

#PBS -N filter_depth.pbs
#PBS -l walltime=00:30:00
#PBS -j oe
#PBS -l select=1:ncpus=1:mem=4gb
#PBS -o /srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/output/filter_depth_log

cd /srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/output

time=$(date "+%Y.%m.%d-%H.%M.%S")

dirname=filterdepth.$time

mkdir $dirname

# get bed file
egrep -v "^#" /srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/data/2020-09-22.ReferenceGenomes/e_coli/busco3/run_e_coli/full_table_e_coli.tsv | \
cut -f3,4,5 | \
sed -e 's/\t/ /g' > $dirname/e_coli.bed | \
egrep -v '^$' > $dirname/e_coli.bed | \
sort -n -k 2 $dirname/e_coli.bed > $dirname/e_coli.sorted.bed

module load samtools

cp /srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/data/2020-09-22.ReferenceGenomes/e_coli/bam/e_coli.bam $dirname/

# filter out reads shorter than a certain value with awk?
samtools view -h $dirname/e_coli.bam | awk -v fil="${FILTER_LEN}" 'length($10) > fil || $1 ~ /^@/' | samtools view -bS - > $dirname/e_coli.filtered.bam


samtools sort -o $dirname/e_coli.sort.bam $dirname/e_coli.filtered.bam
samtools index -b $dirname/e_coli.sort.bam
cp $dirname/e_coli.sort.bam.bai $dirname/e_coli.filtered.bam.bai
samtools mpileup -BQ0 -l $dirname/e_coli.sorted.bed $dirname/e_coli.filtered.bam > $dirname/pileup.out


# samtools view -h ${OD}/${NAME}.bam | awk "length($10) > ${FILTER_LEN} || $1 ~ /^@/" | samtools view -bS - > ${OD}/${NAME}_filtered.bam

# samtools sort -@ ${THREADS} -o ${OD}/${NAME}.${FILTER_LEN}.filtered.sort.bam ${OD}/${NAME}.${FILTER_LEN}.filtered.bam
# samtools index -@ ${THREADS} -b ${OD}/${NAME}.${FILTER_LEN}.filtered.sort.bam
# cp ${OD}/${NAME}.${FILTER_LEN}.filtered.sort.bam.bai ${OD}/${NAME}.${FILTER_LEN}.filtered.bam.bai

# samtools mpileup -BQ0 -l ${OD}/${NAME}.${FILTER_LEN}.sorted.bed ${OD}/${NAME}.${FILTER_LEN}.filtered.bam > ${PILEUP}