#!/bin/bash

#PBS -N run_genomeSize.py
#PBS -l select=2:ncpus=1:mem=4gb
#PBS -l walltime=00:30:00
#PBS -M z5215600@unsw.edu.au
#PBS -m ae  
#PBS -j oe
#PBS -o /srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/output/chelsea_test

# set -e						# if any error occurs, exit 1

##############
## UNTESTED ##
##############

# BAM=/srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/data/2020-09-22.ReferenceGenomes/${NAME}/bam/${NAME}.bam
# OD=/srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/output/chelsea_test
# FILTER_LEN=5000
# SCO=/srv/scratch/z3452659/BINF6112-Sep20/TeamGenomeSize/data/2020-09-22.ReferenceGenomes/${NAME}/busco3/run_${NAME}/full_table_${NAME}.tsv
# WD=$(pwd)
# NAME=${NAME}

#########
# test runs
# qsub -v bam=${BAM},wd=${WD},od=${OD},sco=${SCO},name=${NAME},\
#  filter_len=${FILTER_LEN},assumptions=${ASSUMPTIONS} run.pbs
cd $wd

##################
## LOAD MODULES ##
##################
module load samtools/1.10


##################
## EXPAND FLAGS ##
##################
THREADS=2
FILTER_LEN=${filter_len}
BAM=${bam}
WD=${wd}
OD=${od}
SCO=${sco}
NAME=${name}
ASSUMPTIONS=${assumptions}
VOLUME=${OD}/${NAME}_read_volume.txt
PILEUP=${OD}/${NAME}_pileup.out
METHOD=$( echo ${ASSUMPTIONS} | cut -d',' -f1 | cut -d'=' -f2 )
INDEL=$( echo ${ASSUMPTIONS} | cut -d',' -f2 | cut -d'=' -f2 )
R_CLIPPING=$( echo ${ASSUMPTIONS} | cut -d',' -f3 | cut -d'=' -f2 )

#################
## READ VOLUME ##
#################

# calculate samtools stats to get one volume value
if [[ -z "${FILTER_LEN}" ]]
then
    samtools stats -@ ${THREADS} ${BAM} | grep ^RL | cut -f 2- | awk '{total+=$1*$2} END{print total}' > ${VOLUME}
else
    samtools stats -@ ${THREADS} ${BAM} | grep ^RL | cut -f 2- | awk -v min="${FILTER_LEN}" '{if($1 >= min) total+=$1*$2 } END{print total}' > ${VOLUME}
fi


##################
## READ PILEUPS ##
##################

# get bed file
egrep -v "^#" ${sco} | cut -f3,4,5 | sed -e 's/\t/ /g' > ${OD}/${NAME}.bed
egrep -v '^$' > ${OD}/${NAME}.bed
# create sorted bam file
sort -n -k 2 ${OD}/${NAME}.bed > ${OD}/${NAME}_sorted.bed

cp ${bam} ${OD}/
samtools sort -@ ${THREADS} -o ${OD}/${NAME}.sort.bam ${OD}/${NAME}.bam
samtools index -@ ${THREADS} -b ${OD}/${NAME}.sort.bam
cp ${OD}/${NAME}.sort.bam.bai ${OD}/${NAME}.bam.bai

samtools mpileup -l ${OD}/${NAME}_sorted.bed ${OD}/${NAME}.bam > ${PILEUP}


#################
## GENOME SIZE ##
#################

echo "about to run genomeSize.py"

echo OD=${OD}
echo NAME=${NAME}
echo PILEUP=${pileup}
echo VOLUME=${VOLUME}
echo METHOD=${METHOD}
echo INDEL=${INDEL}
echo R_CLIPPING=${R_CLIPPING}

# taking outputs and giving an answer
python3 genomeSize.py -od ${OD} -n ${NAME} -p ${PILEUP} -v ${VOLUME} -m ${METHOD} -i ${INDEL} -rc ${R_CLIPPING}
